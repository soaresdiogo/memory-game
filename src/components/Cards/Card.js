import React from 'react';
import MUCard from '@material-ui/core/Card';
import CardActionArea from '@material-ui/core/CardActionArea';
import CardContent from '@material-ui/core/CardContent';
import CardMedia from '@material-ui/core/CardMedia';
import Typography from '@material-ui/core/Typography';
import { withStyles } from '@material-ui/core/styles';
import classNames from 'classnames';
import PropTypes from 'prop-types';

const width = '160px';
const height = '200px';
const styles = {
  root: {
    width,
    height,
    margin: '12px',
    cursor: 'pointer',
  },
  flipper: {
    transition: '0.6s',
    transformStyle: 'preserve-3d',
    position: 'relative',
  },
  flipperRotate: {
    transform: 'rotateY(180deg)',
  },
  page: {
    width,
    height,
    position: 'absolute',
    top: 0,
    left: 0,
    backfaceVisibility: 'hidden',
  },
  pageFront: {
    backgroundColor: 'silver',
    transform: 'rotateY(0deg)',
    zIndex: 2,
  },
  pageBack: {
    backgroundColor: 'tomato',
    transform: 'rotateY(180deg)',
  },
  media: {
    height: 110,
  },
};

const Card = ({ name, startLetter, image, classes, isActive, onClick }) => (
  <div
    data-testid="card"
    role="presentation"
    className={classes.root}
    onClick={onClick}
  >
    <div
      className={classNames(classes.flipper, {
        [classes.flipperRotate]: isActive,
      })}
    >
      <MUCard className={classNames(classes.page, classes.pageFront)} />
      <MUCard className={classNames(classes.page, classes.pageBack)}>
        <CardActionArea>
          <CardMedia
            className={classes.media}
            image={image}
            title="Contemplative Reptile"
          />
          <CardContent>
            <Typography variant="h5" component="h2" align="center">
              {startLetter}
            </Typography>
            <Typography variant="h5" component="h2" align="center">
              {name}
            </Typography>
          </CardContent>
        </CardActionArea>
      </MUCard>
    </div>
  </div>
);

Card.defaultProp = {
  classes: PropTypes.node,
  onClick: PropTypes.func,
};

Card.propTypes = {
  name: PropTypes.string.isRequired,
  startLetter: PropTypes.string.isRequired,
  image: PropTypes.string.isRequired,
  isActive: PropTypes.bool.isRequired,
};

export default withStyles(styles)(Card);
