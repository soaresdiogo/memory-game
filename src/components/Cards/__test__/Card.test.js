import React from 'react';
import ReactDOM from 'react-dom';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';

import Card from '../Card';

import '@testing-library/jest-dom/extend-expect';

const CardComponent = (
  <Card
    name="Elephant"
    startLetter="E"
    image="/images/elephant.jpg"
    isActive={false}
  />
);

const renderCard = () => {
  const { getByTestId } = render(CardComponent);
  const container = getByTestId('card');
  return {
    getByTestId,
    container,
  };
};

describe('Card component', () => {
  test('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(CardComponent, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test('should be defined', () => {
    const { getByTestId } = renderCard();
    expect(getByTestId('card')).toBeDefined();
  });

  test('matches snapshot', () => {
    const tree = renderer.create(CardComponent).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
