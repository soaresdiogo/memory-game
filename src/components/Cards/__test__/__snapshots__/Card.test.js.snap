// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component matches snapshot 1`] = `
<div
  className="Card-root-15"
  data-testid="card"
  role="presentation"
>
  <div
    className="Card-flipper-16"
  >
    <div
      className="MuiPaper-root MuiCard-root Card-page-18 Card-pageFront-19 MuiPaper-elevation1 MuiPaper-rounded"
    />
    <div
      className="MuiPaper-root MuiCard-root Card-page-18 Card-pageBack-20 MuiPaper-elevation1 MuiPaper-rounded"
    >
      <button
        className="MuiButtonBase-root MuiCardActionArea-root"
        disabled={false}
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <div
          className="MuiCardMedia-root Card-media-21"
          style={
            Object {
              "backgroundImage": "url(\\"/images/elephant.jpg\\")",
            }
          }
          title="Contemplative Reptile"
        />
        <div
          className="MuiCardContent-root"
        >
          <h2
            className="MuiTypography-root MuiTypography-h5 MuiTypography-alignCenter"
          >
            E
          </h2>
          <h2
            className="MuiTypography-root MuiTypography-h5 MuiTypography-alignCenter"
          >
            Elephant
          </h2>
        </div>
        <span
          className="MuiCardActionArea-focusHighlight"
        />
      </button>
    </div>
  </div>
</div>
`;
